src/main/java/com/ldtteam/blockui/controls/ItemIcon.java

```diff
@@ -138,32 +307,166 @@ public List<Component> getModifiedItemStackTooltip()
+            tooltipList.add(wrapShift(Component.translatable("blockui.tooltip.properties"), !blockStateExtension.alwaysAddBlockStateTooltip()));
+
+            final BlockState blockState = blockStateExtension.blockState();
+            for (final Property<?> property : blockState.getProperties())
+            {
+                tooltipList.add(wrapShift(Component.literal("  " + property.getName() + " = " + getValueName(blockState, property))
+                    .withStyle(ChatFormatting.GRAY), !blockStateExtension.alwaysAddBlockStateTooltip()));
+            }
+
+            if (shouldFixPrevTooltipSize)
+            {
+                prevTooltipSize = tooltipList.size();
+            }
+        }
+
+        if (prevTooltipSize != tooltipList.size())
+        {
+            // add "show more info" text
+            tooltipList.add(ToggleableTextComponent.ofNegated(Screen::hasShiftDown, Component.empty()));
+            tooltipList.add(ToggleableTextComponent.ofNegated(Screen::hasShiftDown,
+                Component.translatable("blockui.tooltip.item_additional_info", Component.translatable("key.keyboard.left.shift"))
+                    .withStyle(ChatFormatting.GOLD)));
+        }
+
+        tooltipList.add(nameOffset, FIX_VANILLA_TOOLTIP);
+        return tooltipList;
     }
 
     private static MutableComponent wrapShift(final MutableComponent wrapped)
     {
         return ToggleableTextComponent.of(Screen::hasShiftDown, wrapped);
     }
+
+    private static MutableComponent wrapShift(final MutableComponent wrapped, final boolean shouldWrap)
+    {
+        return shouldWrap ? ToggleableTextComponent.of(Screen::hasShiftDown, wrapped) : wrapped;
+    }
+
+    /**
+     * Extracts blockstate/modeldata from current itemstack
+     */
+    protected void readBlockStateFromCurrentItemStack()
+    {
+        if (!(itemStack.getItem() instanceof final BlockItem blockItem))
+        {
+            return;
+        }
+    
+        BlockState blockstate = blockItem.getBlock().defaultBlockState();
+
+        // parse block state
+        final CompoundTag blockStateTag = itemStack.getTagElement(BlockItem.BLOCK_STATE_TAG);
+
+        if (blockStateTag == null)
+        {
+            return;
+        }
+
+        final StateDefinition<Block, BlockState> statedefinition = blockstate.getBlock().getStateDefinition();
+
+        for (final String propertyKey : blockStateTag.getAllKeys())
+        {
+            final Property<?> property = statedefinition.getProperty(propertyKey);
+            if (property != null)
+            {
+                blockstate = updateState(blockstate, property, blockStateTag.getString(propertyKey));
+            }
+        }
+
+        // try parsing blockentity
+        final CompoundTag blockEntityTag = itemStack.getTagElement(BlockItem.BLOCK_ENTITY_TAG);
+        BlockEntity be = null;
+        if (blockEntityTag != null)
+        {
+            try
+            {
+                // use probably invalid pos
```

Comment:
I would still extract this in a special class, can extend ItemIcon, but I wouldnt conflate the two.

Reply:
The issue is that it would require new key in loader, which disables the ability to parse nbt correctly

Reply:
Sorry, I mean, all the BlockState ItemIcon stuff, I would put into a BlockStateItemIcon class, so we can use them separately. This makes maintaining the logic easier

Reply:
Yeah, but they are logically connected, that's the issue, I can separate them, but the loader of itemicon element needs to create the blockstate version then

Reply:
Cant it be a BlockStateItemIcon extends ItemIcon?

Reply:
Oh, even for the normal ItemIcon then? We cant have two versions for the xml loader?

Reply:
Yep, I will look into it evening

Reply:
Done as requested+said - made it extend itemIcon and xml loads blockstate aware

Reply:
I still don't see why they should be logically connected at all, that's why I made a separate BlockIcon that has nothing to do with ItemStacks.

Reply:
Probably "all" reasons i know about:
There are blocks which dont have models (barrier, light blocks, ...) but have items
You cant get tooltip without itemstack (especially problem when tooltiping DO blocks)
ItemStack itself can be provide gui translations for 3d models (most of normal blockItems) - more compat mod friendly
There are blocks which even dont have translation defined - but as items they have it (just one iirc)

Reply:
> There are blocks which dont have models (barrier, light blocks, ...) but have items

Those would normally not be present in schematics, but even if they are then that's probably not a problem to show an empty icon with tooltip, since that reflects how the block does actually look in-world.

> You cant get tooltip without itemstack (especially problem when tooltiping DO blocks)

That's a feature, not a bug.  The itemstack tooltip is entirely unrelated to the block tooltip, and it should not be assumed that they're the same.

> There are blocks which even dont have translation defined - but as items they have it (just one iirc)

That's a bug in DO.  You can see missing translations for DO blocks in Jade and in F3; I have reported that previously but I was told we didn't care enough to fix it.

> ItemStack itself can be provide gui translations for 3d models (most of normal blockItems) - more compat mod friendly

The rotation of the item model and the rotation of the block model need not be related in any way, and it seems incorrect to assume that they are.

Reply:
1) I'm skipping (false arguments) relations of blockui to structurize just as blueprints to mcol - yes the mods (and their features) are mainly for mcol usage, but that doesn't mean it should not be implemented with respect to all possible situations

2) well partially tru, but vanilla doesn't for obvious reason have tooltips for blockstates, but we want tooltip then we need to get the base somewhere, and the only remaining source of somewhat true data are items. So saying entirely unrelated is wrong, cuz it's only relation remaining (else you can provide other source). Also it should be fairly easy to fix the logic so automatic tooltip update doesn't trigger after manual tooltip override was set

3) it was vanilla block iirc, but there might be more purely blocky blocks (other mods) which aren't supposed to exist elsewhere but chunk data

4) completely wrong, there is ItemDisplayContext.GUI for the exact thing u mentioned - rotating block models for gui